lefthk-core/src/ipc.rs

```diff
@@ -59,33 +59,28 @@ impl Pipe {
         PathBuf::from(format!("command-{}.pipe", display))
     }
 
-    pub async fn read_command(&mut self) -> Option<Command> {
-        self.rx.recv().await
+    pub async fn get_next_command(&mut self) -> Option<Box<dyn Command>> {
+        if let Some(normalized_command) = self.rx.recv().await {
+            return command::denormalize(normalized_command).ok();
+        }
+        None
     }
 }
 
-async fn read_from_pipe(pipe_file: &Path, tx: &mpsc::UnboundedSender<Command>) -> Option<()> {
-    let file = fs::File::open(pipe_file).await.ok()?;
-    let mut lines = BufReader::new(file).lines();
+async fn read_from_pipe<'a>(pipe_file: &Path, tx: &mpsc::UnboundedSender<NormalizedCommand>) {
+    if let Ok(file) = fs::File::open(pipe_file).await {
+        let mut lines = BufReader::new(file).lines();
 
-    while let Some(line) = lines.next_line().await.ok()? {
-        let cmd = match parse_command(&line) {
-            Ok(cmd) => cmd,
-            Err(err) => {
-                tracing::error!("An error occurred while parsing the command: {}", err);
-                return None;
+        while let Ok(line) = lines.next_line().await {
+            if let Some(content) = line {
+                if let Ok(normalized_command) = NormalizedCommand::try_from(content) {
+                    if command::denormalize(normalized_command.clone()).is_ok() {
+                        if let Err(err) = tx.send(normalized_command) {
```

Comment:
I know that you're probably disliking this very much @AethanFoot , but it doesn't look bad in my opinion.